src/datachain/lib/listing.py

```diff
@@ -0,0 +1,106 @@
+import asyncio
+from collections.abc import AsyncIterator, Iterator, Sequence
+from typing import Callable, Optional
+
+from botocore.exceptions import ClientError
+from fsspec.asyn import get_loop
+
+from datachain.asyn import iter_over_async
+from datachain.client import Client
+from datachain.error import ClientError as DataChainClientError
+from datachain.lib.file import File
+
+ResultQueue = asyncio.Queue[Optional[Sequence[File]]]
+
+DELIMITER = "/"  # Path delimiter
+FETCH_WORKERS = 100
+
+
+async def _fetch_dir(client, prefix, result_queue) -> set[str]:
```

Comment:
These private functions were copied from [our fsspec client](https://github.com/iterative/datachain/blob/main/src/datachain/client/fsspec.py) and adapted for usage in listing generator